content/docs/document-lifecycle.md

```diff
@@ -0,0 +1,55 @@
+
+After all pre-processing steps have been completed, the document's state is set to `Processed`.
+
+Please note that resolution of cross-references are not permitted until this point.
+
+## Linking
+Once all pre-processing steps are complete, cross-references are resolved via the `Linker` service. The `Linker` gets all cross-references in a `LangiumDocument` and tries to resolve them. For each cross-reference, the `Linker` tries to find the correct `AstNode` and its location inside of a `LangiumDocument`. The linker uses the scopes computed during the pre-processing to identify the `AstNode` which is the target of a reference. With the default implementation, only reference that targets `AstNode`s registered in the `LangiumDocuments` service can be resolved (i.e. documents present in the current workspace). Linking can resolve references lazily if needed before the first eager resolution. 
+
+Once the linking is complete, the document's state is set to `Linked`.
+
+## Indexing of the cross-references
+Once the cross-references have been resolved by the linker, the document URI and the references it contains are mapped. This ensures an efficient lookup to identify documents that may be impacted by a change in a `LangiumDocument`.
+
+After the cross-references have been indexed, the document's state is set to `IndexedReferences`.
+
+## Validation
+The `DocumentsValidator` creates an array of `Diagnostic` from a `langiumDocument`. This array of `Diagnostics` contains all errors that could have ocurred during lexing, parsing, and linking, and the results of a set of custom validations (i.e. language-specific validations).
+
+After the diagnostics have been created, the document's state is set to `Validated`.
+
+At this point, all documents have been processed by the `DocumentBuilder` and the workspace is initialized.
+
+## Modifications of a document
+When a `TextDocument` is modified, the client notifies the server, which triggers corresponding events. In Langium, a change in a `TextDocument`'s content or location leads to the invalidation of the associated `LangiumDocument`. The document's state is set to `Changed` and the document's entry is removed from the `documentMap` in the `LangiumDocuments` service. If the `TextDocument` was deleted, the corresponding `langiumDocument` is removed from the index in the `IndexManager` service. If the document's content or URI was modified, a new instance of `LangiumDocument` is created as described [here](#creation-of-langiumdocuments).  All other documents that may have been affected as a result of the modification get their references unlinked and their state set to the lowest from their own state or `Processed` (i.e. before linking).
```

Comment:
`documentMap` is an implementation detail of `LangiumDocuments` and should not be mentioned here.